### ISSUE TITLE
Date timestamp attr

### ISSUE THREAD

**Issue by qooob** [webmin/webmin#820]
Jamie, hi.

Is this possible to just have `data-timestamp` and `data-timestamp-type` attr on the tags with the data/time?

I need to convert the strings right, so it would look the same across all over.
https://github.com/qooob/authentic-theme/issues/1013

For example, at Recent Logins on the containing tag, which contain the text string (td), you could simply go and add an attr with `data-timestamp="1518737140"` and with another attr (same td tag in this example) type `data-timestamp-type` could equal to `fromNow`. Using _momentjs_, which is very powerful, we could deliver any date/time format based on preferences (from theme configuraion for now) to the user.

Besides `fromNow`, `data-timestamp-type` values could also be `full` (default and could be dropped) and `short` for short date/time format.

Can we start from Recent Logins and then later will be able to have it all over? _Webmin Configuration/User interface options/Format for displayed dates_ has very limited options. 

In case you support this idea, I will move date/time configurations to be displayed at _User interface options/Format for displayed dates_, away from Theme Configuration.

**Comment by jcameron** [comment]
In Webmin, all unix time to human readable data/time conversions should be done by the `make_date` function on the server side. There's a setting somewhere in the Webmin Configuration module to control which date format it uses.

**Comment by qooob** [comment]
Yes, this is actually what I was saying, but it's very poor compare to _momentjs_.


You mean this, right?
![screenshot from 2018-02-18 23-47-03](https://user-images.githubusercontent.com/4426533/36356680-3878e762-1506-11e8-9275-adf9a74b5b1d.png)

**Comment by jcameron** [comment]
Yes. I'm open to adding more options, or even changing the format.

**Comment by qooob** [comment]
Could you do it for Recent Logins for the start?

**Comment by jcameron** [comment]
Does that page not use the make_date function to format the timestamp?

**Comment by qooob** [comment]
It has very limited amount of date formats, it doesn't have short date format, it doesn't have _ago_ format.

Most importantly it binds user to certain format without letting set your own one.

Right now in the theme you can set any format you want from its configuration.


MomentJS supports any format and has Multilingual support (formatting dates based in selected language - this is also supported in theme already).

**Multiple Locale Support**
```
moment.locale();         // en
moment().format('LT');   // 10:55 AM
moment().format('LTS');  // 10:55:30 AM
moment().format('L');    // 02/21/2018
moment().format('l');    // 2/21/2018
moment().format('LL');   // February 21, 2018
moment().format('ll');   // Feb 21, 2018
moment().format('LLL');  // February 21, 2018 10:55 AM
moment().format('lll');  // Feb 21, 2018 10:55 AM
moment().format('LLLL'); // Wednesday, February 21, 2018 10:55 AM
moment().format('llll'); // Wed, Feb 21, 2018 10:55 AM
```

**Relative Time**

```
moment("20111031", "YYYYMMDD").fromNow(); // 6 years ago
moment("20120620", "YYYYMMDD").fromNow(); // 6 years ago
moment().startOf('day').fromNow();        // 11 hours ago
moment().endOf('day').fromNow();          // in 13 hours
moment().startOf('hour').fromNow();       // an hour ago
```

Ability to **Format Dates** at will:

```
moment().format('MMMM Do YYYY, h:mm:ss a'); // February 21st 2018, 10:56:22 am
moment().format('dddd');                    // Wednesday
moment().format("MMM Do YY");               // Feb 21st 18
moment().format('YYYY [escaped] YYYY');     // 2018 escaped 2018
moment().format();                          // 2018-02-21T10:56:22+03:00
```

All I ask is one `data-attr` on the container field that would export Unix Timestamp.

**Comment by qooob** [comment]
What do you think, Jamie?

**Comment by jcameron** [comment]
What if we allowed the theme to override the `make_date` function?

**Comment by qooob** [comment]
Oh, that would be great!

**Comment by jcameron** [comment]
Ok, just checked in support for this.

**Comment by qooob** [comment]
Okay, logic works perfect. How can I pass on the return `<span data-1 data-2 data-3></span>`? It seems that Webmin core removes HTML completely?

**Comment by qooob** [comment]
Ah, okay. It doesn't filter out HTML, it just breaks the string on space, in case time is present. Okay, I'll figure this out from here. I will make a commit tomorrow to support unified date/time format across Webmin.

Once question, `$gconfig{'dateformat'}` is only set by _root_  user in Webmin Configuration?

**Comment by qooob** [comment]
Is this okay, if I let any other user, using theme's settings (in user-mode) to define date format manually, in case he/she would want something that wouldn't blend with default/language type of format?

**Comment by qooob** [comment]
It would be nice but difficult to make. For now, we work with HTML output. It's less flexible and possibly a bit slower but works pretty much the same way to the end user.

Having mentioned API, yes, it would be easier to creat mobile apps. We will get there one day, I believe.

**Comment by chris001** [comment]
Yes, an API would surely raise the prominence of the project substantially.

**Comment by jcameron** [comment]
@iliaross - I'm OK if there is a theme option to change the output of this function.

**Comment by iliaross** [comment]
❗️ If the issue isn't resolved, feel free to reopen it with updated details.
